cpp/src/layout/barnes_hut.h

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph.h>
+#include <internals.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm_utils.h>
+#include <stdio.h>
+#include <converters/COOtoCSR.cuh>
+#include <graph.hpp>
+#include <rmm/device_buffer.hpp>
+
+#include "bh_kernels.h"
+#include "fa2_kernels.h"
+#include "utilities/error_utils.h"
+#include "utilities/graph_utils.cuh"
+#include "utils.h"
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+void barnes_hut(experimental::GraphCOOView<vertex_t, edge_t, weight_t> &graph,
+                float *pos,
+                const int max_iter                            = 1000,
+                float *x_start                                = nullptr,
+                float *y_start                                = nullptr,
+                bool outbound_attraction_distribution         = true,
+                bool lin_log_mode                             = false,
+                bool prevent_overlapping                      = false,
+                const float edge_weight_influence             = 1.0,
+                const float jitter_tolerance                  = 1.0,
+                const float theta                             = 0.5,
+                const float scaling_ratio                     = 2.0,
+                bool strong_gravity_mode                      = false,
+                const float gravity                           = 1.0,
+                bool verbose                                  = false,
+                internals::GraphBasedDimRedCallback *callback = nullptr)
+{
+  const edge_t e   = graph.number_of_edges;
+  const vertex_t n = graph.number_of_vertices;
```

Comment:
Would all of these functions work fine if you pass a zero edge graph? Is there any sort of checking required to avoid zero block cuda kernel launches?

Reply:
So edges are used only for the attraction kernel and indeed having 0 edge graph results in a division by zero when computing block dimensions. Added a check for that so we avoid launching the kernel at all.